core/src/main/java/com/amplifyframework/analytics/UserProfile.java

```diff
@@ -0,0 +1,313 @@
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.analytics;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import java.util.Objects;
+
+/**
+ * Represents user specific data such as name, email, plan, location etc.
+ */
+public final class UserProfile {
+    private String name;
+    private String email;
+    private String plan;
+    private Location location;
+    private Properties customProperties;
+
+    public UserProfile(Builder builder) {
+        this.name = builder.name;
+        this.email = builder.email;
+        this.plan = builder.plan;
+        this.location = builder.location;
+        this.customProperties = builder.customProperties;
+    }
+
+    @Nullable
+    public String getName() {
+        return name;
+    }
+
+    @Nullable
+    public String getEmail() {
+        return email;
+    }
+
+    @Nullable
+    public String getPlan() {
+        return plan;
+    }
+
+    @Nullable
+    public Location getLocation() {
+        return location;
+    }
+
+    @Nullable
+    public Properties getCustomProperties() {
+        return customProperties;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @SuppressWarnings("checkstyle:NeedBraces")
+    @Override
+    public boolean equals(Object object) {
+        if (this == object) return true;
+        if (object == null || getClass() != object.getClass()) return false;
+
+        UserProfile that = (UserProfile) object;
+
+        if (name != null ? !name.equals(that.name) : that.name != null) return false;
+        if (email != null ? !email.equals(that.email) : that.email != null) return false;
+        if (plan != null ? !plan.equals(that.plan) : that.plan != null) return false;
+        if (location != null ? !location.equals(that.location) : that.location != null)
+            return false;
+        return customProperties != null ? customProperties.equals(that.customProperties) :
+                that.customProperties == null;
+    }
+
+    @SuppressWarnings("checkstyle:MagicNumber")
+    @Override
+    public int hashCode() {
+        int result = name != null ? name.hashCode() : 0;
+        result = 31 * result + (email != null ? email.hashCode() : 0);
+        result = 31 * result + (plan != null ? plan.hashCode() : 0);
+        result = 31 * result + (location != null ? location.hashCode() : 0);
+        result = 31 * result + (customProperties != null ? customProperties.hashCode() : 0);
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "UserProfile{" +
+                "name='" + name + '\'' +
+                ", email='" + email + '\'' +
+                ", plan='" + plan + '\'' +
+                ", location=" + location +
+                ", customProperties=" + customProperties +
+                '}';
+    }
+
+    public static final class Builder {
```

Comment:
These Builder methods all should return `@NonNull`.

You can configure your IDE to warn when things don't have these annotations, actually. (I have this setup.)

Our code base is too far "off" right now to enable it with an automated tool at compile time. But it's still possible to get the suggestions in the IDE.